sockeye/beam_search.py

```diff
@@ -670,7 +678,45 @@ def forward(self,
                 if t == 1:
                     scores *= first_step_mask
 
-                best_hyp_indices, best_word_indices, scores_accumulated = self._top(scores, offset)
+                vocab_size = scores.shape[1]
+
+                # Groupwise shape: (batch_size, num_diverse_groups, diverse_group_size[, vocab_size])
```

Comment:
Instead of having this entire block inside the beam search class, would it be possible to move it into a separate Block (I assume not HybridBlock as you are using numpy-style indexing) that replaces `TopK`? The reason is that the Python for loop and the advanced indexing operations are likely slow.
Separating this code into its own block would be a bit cleaner I think.

Reply:
Maybe something like `GroupedTopK` or `DiverseTopK`

Reply:
Ok, thanks! Moved it to a separate `Block`.

I'm having a hard time figuring out how to avoid the imperative `for` loop and indexing, but if you have a hint or suggestion, I'll be happy to have a stab at it.